File: javascript/merge_lists.js
/*
 Merge two sorted lists together so the result is also sorted
 Duplicates are allowed
 */

class Solution {
    solve(a, b) {
        let i = 0;
        let y = 0;
        let result = [];

        while (i < a.length && y < b.length) {
            if (a[i] <= b[y]) {
                result.push(a[i]);
                i++;
            } else {
                result.push(b[y]);
                y++;
            }
        }

        while (i < a.length) {
            result.push(a[i]);
            i++;
        }

        while (y < b.length) {
            result.push(b[y]);
            y++;
        }

        return result;
    }
}

let solution = new Solution;
let a = [1, 3, 5, 6];
let b = [1, 4, 5, 7, 8];

console.log(solution.solve(a, b));

File: swift/merge_lists.swift
/*
 Merge two sorted lists together so the result is also sorted
 Duplicates are allowed
*/

class Solution {
  func merge(a: [Int], b: [Int]) -> [Int] {
    var ai = 0, bi = 0, result = [Int]()

    while ai < a.count && bi < b.count {
      if a[ai] < b[bi] {
        result.append(a[ai])
        ai += 1
      } else {
        result.append(b[bi])
        bi += 1
      }
    }

    while ai < a.count {
      result.append(a[ai])
      ai += 1
    }

    while bi < b.count {
      result.append(b[bi])
      bi += 1
    }

    return result
  }
}

let s = Solution()
print(s.merge(a: [1, 3, 4, 5, 6], b: [2, 6, 7]))
